File: src/server/GameServer.java
package server;

import client.GameClient;
import engine.controller.GameTypeHandler;
import engine.controller.RespawnController;
import engine.physics.DeltaTime;
import engine.physics.PhysicsController;
import engine.entities.Player;
import javafx.geometry.Point2D;
import networking.server.ConnectedPlayers;
import javafx.scene.transform.Rotate;
import networking.server.ServerNetwork;
import engine.controller.PowerUpController;

public class GameServer extends Thread {

	private ServerGameState gameState;
	private ServerNetwork network;

	private boolean running;

	public GameServer(ServerGameState gameState) {
		this.gameState = gameState;
		this.network = new ServerNetwork(this.gameState);
	}

	public void run() {
		this.network.start();

		PowerUpController puController = new PowerUpController(gameState, this.network.getDispatcher());
		RespawnController resController = new RespawnController(gameState, this.network.getDispatcher());
		PhysicsController physicsController = new PhysicsController(gameState,this.network.getDispatcher());

        this.running = true;
        long lastTime = System.nanoTime();
		while (this.running) {
			// Server logic

			if (!this.gameState.isStarted()) {

				try {
					Thread.sleep(1);
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
				this.waitForPlayersToConnect();

			} else {

				physicsController.serverLoop();

				puController.serverUpdate();

				resController.update();

				this.running = GameTypeHandler.checkRunning(gameState);

				Thread.yield();
				this.sendLocations();
				this.sendHealthUpdates();

				GameClient.timeElapsed += DeltaTime.deltaTime;
                //calculate deltaTime
                long time = System.nanoTime();
                DeltaTime.deltaTime = (int) ((time - lastTime) / 1000000);
                lastTime = time;

				try {
					Thread.sleep(15); // Every second
				} catch (InterruptedException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
			}
		}

		this.network.getDispatcher().broadcastGameEnded();

		this.network.close();
	}

	private void waitForPlayersToConnect() {
		ConnectedPlayers connectedPlayers = this.network.getConnectedPlayers();
		//todo can't get the connected player : how and where player added

        // once players are all connected, start the game
        if(connectedPlayers != null && connectedPlayers.count() == gameState.getNumPlayers()) {
			// Start the game
			connectedPlayers.assignStartingLocations(gameState.getMap().getWidth(), gameState.getMap().getHeight());
			this.gameState.getScoreBoard().initialise(this.gameState.getAllPlayers());
			this.network.getDispatcher().broadcastGameState();

			if (connectedPlayers.allHaveInitialGameState()) {
				this.gameState.setStarted(true);
				this.network.getDispatcher().broadcastGameStarted();
				GameClient.timeElapsed = 0;

			}
		}
	}

	private void sendLocations() {
		synchronized (gameState.getAllPlayers()) {
			for (Player p : gameState.getAllPlayers()) {
				if (p.isAlive()) {
					Point2D location = p.getLocation();
					Rotate playerAngle = p.getPlayerAngle();
					double x = location.getX();
					double y = location.getY();

					this.network.getDispatcher().broadcastLocationState(p.getId(), x, y, playerAngle.getAngle());
				}
			}
		}
	}

	private void sendHealthUpdates() {
		synchronized (gameState.getAllPlayers()) {
			for (Player p : gameState.getAllPlayers()) {
				float playerHealth = p.getHealth();

				this.network.getDispatcher().broadcastHealthState(p.getId(),playerHealth);
			}
		}
	}

	public ServerNetwork getNetwork() {
		return network;
	}
}

File: neon-element-cpp/src/server/GameServer.cpp
#include "GameServer.hpp"
#include <chrono>
#include <thread>

namespace neon {

GameServer::GameServer(ServerGameState& state)
    : state_(state)
    , physics_(state)
    , puController_(state)
    , resController_(state)
{
    network_ = std::make_unique<net::ServerNetwork>(state_);
}

GameServer::~GameServer()
{
    stop();
}

void GameServer::start()
{
    if (running_.exchange(true)) return;
    thread_ = std::jthread([this](std::stop_token) { run(); });
}

void GameServer::stop()
{
    running_.store(false);
    if (thread_.joinable()) {
        thread_.request_stop();
        thread_.join();
    }
    network_->stop();
}

void GameServer::run()
{
    // Phase 1: wait for all players to connect
    waitForPlayersToConnect();

    if (!running_.load()) return;

    // Phase 2: game loop
    gameLoop();

    // Game over
    network_->getDispatcher().broadcastGameEnded();
    network_->stop();
}

void GameServer::waitForPlayersToConnect()
{
    auto& connPlayers = network_->getConnectedPlayers();

    while (running_.load()) {
        if (connPlayers.count() >= state_.getNumPlayers()) {
            // All players connected
            connPlayers.assignStartingLocations(
                state_.getMap().getWidth(),
                state_.getMap().getHeight());

            state_.rebuildPlayerList();
            state_.getScoreBoard().initialise(state_.getAllPlayers());

            // Broadcast initial game state
            network_->getDispatcher().broadcastGameState();

            // Wait for all clients to ACK
            while (running_.load() && !connPlayers.allHaveInitialGameState()) {
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
            }

            if (!running_.load()) return;

            // Start the game
            state_.setStarted(true);
            state_.start();
            network_->getDispatcher().broadcastGameStarted();
            clock_.start();
            return;
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
}

void GameServer::gameLoop()
{
    using namespace std::chrono;

    while (running_.load()) {
        clock_.update();

        physics_.serverLoop(clock_);
        puController_.update(clock_);
        resController_.update(clock_);

        if (!GameTypeHandler::checkRunning(state_, clock_)) {
            state_.stop();
            break;
        }

        sendLocations();
        sendHealthUpdates();

        std::this_thread::sleep_for(milliseconds(15));
    }
}

void GameServer::sendLocations()
{
    auto& dispatcher = network_->getDispatcher();
    auto& players = state_.getAllPlayers();

    for (auto* player : players) {
        if (player->isAlive()) {
            auto loc = player->getLocation();
            dispatcher.broadcastLocationState(
                player->getId(),
                static_cast<double>(loc.x),
                static_cast<double>(loc.y),
                static_cast<double>(player->getAngle()));
        }
    }
}

void GameServer::sendHealthUpdates()
{
    auto& dispatcher = network_->getDispatcher();
    auto& players = state_.getAllPlayers();

    for (auto* player : players) {
        dispatcher.broadcastHealthState(player->getId(), player->getHealth());
    }
}

} // namespace neon
